Serialize and Deserialize Debug Scopes

usama54321 opened the pull request:
Serialize functions referred only from debug scopes as zombies. Serialize rest of the functions normally.

While deserializing, if isDebug (deserializing for debug scopes) only deserialize the declaration as zombie and mark function as private/shared since public functions make into the final binary causing linking errors.

If function is later referred to in non debug scope context and the function was not marked a zombie while serializing, resurrect the function, deserialize body if applicable and return it.

While serializing instructions, whenever the debug scopes of the instruction changes, the new scope along with parents is serialized in post order.

<!--
If this pull request is targeting a release branch, please fill out the
following form:
https://github.com/swiftlang/.github/blob/main/PULL_REQUEST_TEMPLATE/release.md?plain=1

Otherwise, replace this comment with a description of your changes and
rationale. Provide links to external references/discussions if appropriate.
If this pull request resolves any GitHub issues, link them like so:

  Resolves <link to issue>, resolves <link to another issue>.

For more information about linking a pull request to an issue, see:
https://docs.github.com/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

usama54321 commented on lib/FrontendTool/TBD.cpp:
```diff
@@ -85,6 +85,10 @@ static bool validateSymbols(DiagnosticEngine &diags,
       auto externallyVisible =
           (GV->hasExternalLinkage() || GV->hasCommonLinkage())
         && !GV->hasHiddenVisibility();
+      if (auto G = GV->getAliaseeObject())
```
This is no longer needed

usama54321 commented on lib/SIL/IR/SILFunction.cpp:
```diff
@@ -159,10 +159,45 @@ SILFunction *SILFunction::create(
   // Get a StringMapEntry for the function.  As a sop to error cases,
   // allow the name to have an empty string.
   llvm::StringMapEntry<SILFunction*> *entry = nullptr;
+  if (isZombie) {
```
I can create a separate helper function for this instead of changing the existing constructor

eeckstein commented on lib/SIL/Verifier/SILVerifier.cpp:
```diff
@@ -7201,7 +7201,7 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
     case SILLinkage::Package:
     case SILLinkage::Shared:
-      require(F->isDefinition() || F->hasForeignBody(),
+      require(F->isDefinition() || F->hasForeignBody() || F->isZombie(),
```
Zombie functions should not be verified at all. For most parts of the compiler - including the verifier - they don't exist.

usama54321 replied:
Thanks. Let me see why I was hitting this failure.

eeckstein commented on lib/Serialization/SILFormat.h:
```diff
@@ -312,6 +356,7 @@ namespace sil_block {
                      BCFixed<1>,  // is distributed
                      BCFixed<1>,  // is runtime accessible
                      BCFixed<1>,  // are lexical lifetimes force-enabled
+                     BCFixed<1>,  // zombie
```
I would not serialize this flag. Instead just serialize zombie functions as regular functions.

However, if you need this flag to disable linkage verification of such functions, it's better to call it something like "onlyReferencedByDebugInfo".

usama54321 replied:
Sometimes, functions that I serialize just for debug scopes get deserialized by the compiler in a non debug scopes context, e.g. by the generic specializer. In those cases, I try to resurrect the function which causes verification failures because the compiler is expecting the function to have a body which it does not. I guess i can get around that by just not resurrecting declarations that break these invariants, e.g. shared functions that are declarations only etc.. I will revisit this once I have made more progress on the PR

eeckstein replied:
I think what you should do is to _not_ try to de-serialize a "onlyReferencedByDebugInfo" function if it's created from an optimization pass, like the specializer. Instead just create a new function in this case.
Only when de-serializing debug info and a function is referenced from debug info and it doesn't exist, yet (either as a "real" function or as a zombie function), then de-serialize a "onlyReferencedByDebugInfo" function.

usama54321 replied:
Yes that is what I am doing right now here. I just meant to say that maybe I can do this without storing the extra onlyReferencedByDebugInfo bit https://github.com/swiftlang/swift/blob/07401c0b474eeeaec4c9159f4cfd0d015d7f1fb6/lib/Serialization/DeserializeSIL.cpp#L780

usama54321 replied:
I just return null in which case I think the optimizer handles creation of a new function

usama54321 commented on include/swift/SIL/SILFunction.h:
```diff
@@ -522,8 +522,9 @@ class SILFunction
          EffectsKind EffectsKindAttr = EffectsKind::Unspecified,
          SILFunction *InsertBefore = nullptr,
-         const SILDebugScope *DebugScope = nullptr);
+         const SILDebugScope *DebugScope = nullptr, bool isZombie = false);
```
Yes I was planning to remove this. The current changes already are not using this

usama54321 commented on lib/SIL/IR/SILModule.cpp:
```diff
@@ -405,7 +406,7 @@ SILFunction *SILModule::removeFromZombieList(StringRef Name) {
 
 /// Erase a function from the module.
-void SILModule::eraseFunction(SILFunction *F) {
+void SILModule::eraseFunction(SILFunction *F, bool Remove) {
```
Thanks. I will be removing this along with the above change in SILFunction::create

usama54321 commented on lib/SIL/IR/SILFunction.cpp:
```diff
@@ -231,6 +266,38 @@ SILFunction::SILFunction(
     initFunction({this}, &libswiftSpecificData, sizeof(libswiftSpecificData));
 }
 
+SILFunction *SILFunction::resurrectFunction(SILModule &M, StringRef name) {
```
The issue there is it expects the complete list of arguments in the Function::create and calls init with those arguments which made the API very clunky to reuse just for resurrection. Plus the Function builder right now passes defaults while creating a declaration which means I have to reinitialize everything in the resurrected function.

eeckstein replied:
Still, it should never be needed to explicitly resurrect a function (see my other comment).

usama54321 commented on lib/SIL/Parser/ParseSIL.cpp:
```diff
@@ -125,6 +125,13 @@ ParseSILModuleRequest::evaluate(Evaluator &evaluator,
     return SILModule::createEmptyModule(desc.context, desc.conv, desc.opts);
   }
+
+  for (auto Fn : parserState.PotentialZombieFns) {
```
The destructor is getting called too late. I will refactor this into a separate function

usama54321 commented:
@swift-ci test

usama54321 commented:
@swift-ci test

usama54321 commented on lib/Serialization/DeserializeSIL.cpp:
```diff
@@ -532,10 +623,19 @@ SILDeserializer::readSILFunctionChecked(DeclID FID, SILFunction *existingFn,
   PrettyStackTraceStringAction trace("deserializing SIL function", name);
 
-  auto &cacheEntry = Funcs[FID-1];
+  ModuleFile::PartiallySerialized<SILFunction *> *cacheEntryOpt = nullptr;
```
Oops this sneaked in. Will fix

eeckstein commented on lib/Serialization/DeserializeSIL.cpp:
```diff
@@ -667,6 +777,17 @@ SILDeserializer::readSILFunctionChecked(DeclID FID, SILFunction *existingFn,
   if (fn) {
+    assert(!forDebugScope);
+
+    if (fn->isZombie()) {
```
Can this branch ever be taken? `fn` shouldn't be a zombie:
* An `existingFn` which is passed to readSILFunctionChecked should either be null or an existing (non-zombie) declaration.
* `SILMod.lookUpFunction(name)` should not return a zombie

usama54321 replied:
Aha that makes sense. I think this was taken before I started maintaining the invariant at line 740. I will remove this code and add additional assertions

https://github.com/swiftlang/swift/pull/76934/files#diff-a58e4a75394b751f1446f32a43586ebc1356bf6f94c87affe3e9fb0523287366R740

usama54321 committed:
Push assertions in the deserializer

usama54321 committed:
Cleanup code

usama54321 committed:
minor bug fixes

usama54321 commented on test/Serialization/debug-scope.swift:
```diff
@@ -0,0 +1,26 @@
+// CHECK: sil_scope [[SCOPE_GENERIC_THREE:[0-9]+]] { loc "SOURCE_DIR/test/Serialization/Inputs/def_debug.swift":21:12 parent [[SCOPE_GENERIC_TWO]] inlined_at [[INLINE_SITE_TWO]] }
+// CHECK: sil_scope [[SCOPE_GENERIC_FOUR:[0-9]+]] { loc "SOURCE_DIR/stdlib/public/core/Hasher.swift":{{[0-9]+}}:{{[0-9]+}} parent @$ss6HasherV7combineyyxSHRzlFSi_Tgq5 : $@convention(method) (Int, @inout Hasher) -> () inlined_at [[SCOPE_GENERIC_THREE]] }
+// CHECK: sil_scope [[SCOPE_GENERIC_FIVE:[0-9]+]] {  parent [[SCOPE_GENERIC_FOUR]] inlined_at [[SCOPE_GENERIC_THREE]] }
+// CHECK: sil_scope [[SCOPE_GENERIC_SIX:[0-9]+]] { loc "BUILD_DIR/stdlib/public/core/8/IntegerTypes.swift":[[ROW_GENERIC:[0-9]+]]:[[COL_GENERIC:[0-9]+]] parent @$sSiSHsSH4hash4intoys6HasherVz_tFTW : $@convention(witness_method: Hashable) (@inout Hasher, @in_guaranteed Int) -> () inlined_at [[SCOPE_GENERIC_FIVE]] }
```
These lined need to be fixed as the stdlib is not compiled with the new debug flag and as a result these scopes are missing during deserialization

adrian-prantl commented on include/swift/Frontend/FrontendOptions.h:
```diff
@@ -125,6 +125,7 @@ class FrontendOptions {
   /// Include local definitions/references in the index data.
   bool IndexIncludeLocals = false;
 
+  bool SerializeDebugInfo = false;
```
I think this should get a more specific name: SerializeDebugInfoInSIL or something like that.

adrian-prantl commented on include/swift/Serialization/SerializationOptions.h:
```diff
@@ -151,6 +151,7 @@ namespace swift {
 
     bool AutolinkForceLoad = false;
     bool SerializeAllSIL = false;
+    bool SerializeDebugInfo = false;
```
same here

adrian-prantl commented on include/swift/Option/Options.td:
```diff
@@ -805,6 +805,10 @@ def experimental_package_interface_load:
+def experimental_serialize_debug_info:
+  Flag<["-"], "experimental-serialize-debug-info">,
+  Flags<[FrontendOption, HelpHidden]>,
+  HelpText<"Enables seriailzation/deserialization of debug scopes">;
```
This should become on  by default eventually, right?

usama54321 replied:
Yes it should be. I am not sure if we want to turn it on right now. @kubamracek Do you have a suggestion here? Thanks

kubamracek replied:
Yeah, I think this is a staging flag, correct? Sounds like a good approach to me.

adrian-prantl commented on lib/SIL/Parser/SILParserState.h:
```diff
@@ -62,6 +62,8 @@ class SILParserState : public SILParserStateBase {
   bool parseSILProperty(Parser &P) override;
   bool parseSILScope(Parser &P) override;
+
+  void markZombies();
```
Add a doxygen comment explaining what this does?

adrian-prantl commented on lib/Serialization/DeserializeSIL.cpp:
```diff
@@ -379,6 +379,95 @@ SILType SILDeserializer::getSILType(Type Ty, SILValueCategory Category,
+  llvm::Expected<llvm::BitstreamEntry> maybeEntry =
+      SILCursor.advance(AF_DontPopBlockAtEnd);
+  if (!maybeEntry)
+    MF->fatal(maybeEntry.takeError());
```
I don't know if the rest of the deserializer also works this way, but I hope this doesn't call `exit()`. The compiler is also embedded in some long-lived processes (for example lldb) and terminating the process is not a great experience there. If possible returning (and bubbling up) an `llvm::Expected<unsigned>` is much better.

usama54321 replied:
I have bubbled up the error. Calling fatal seems to be a pretty common pattern in the deserializer.

adrian-prantl commented on lib/Serialization/DeserializeSIL.cpp:
```diff
@@ -379,6 +379,95 @@ SILType SILDeserializer::getSILType(Type Ty, SILValueCategory Category,
+      Parent = getFuncForReference(MF->getIdentifierText(ParentID), true);
+      assert(Parent);
+    } else {
+      assert(ParsedScopes.find(ParentID) != ParsedScopes.end());
```
this should also be error handling returning an llvm::Error<>.

adrian-prantl commented on lib/Serialization/DeserializeSIL.cpp:
```diff
@@ -877,8 +994,10 @@ SILDeserializer::readSILFunctionChecked(DeclID FID, SILFunction *existingFn,
 
   GenericEnvironment *genericEnv = nullptr;
-  if (!declarationOnly)
+  //generic signatures are stored for declarations as well in a debug context
```
```suggestion
  // Generic signatures are stored for declarations as well in a debug context.
```

adrian-prantl commented on lib/Serialization/DeserializeSIL.cpp:
```diff
@@ -877,8 +994,10 @@ SILDeserializer::readSILFunctionChecked(DeclID FID, SILFunction *existingFn,
+  //generic signatures are stored for declarations as well in a debug context
+  if (!declarationOnly || onlyReferencedByDebugInfo) {
     genericEnv = MF->getGenericSignature(genericSigID).getGenericEnvironment();
+  }
```
llvm coding style doesn't wrap single statements in `{ }`

adrian-prantl commented on lib/Serialization/SILFormat.h:
```diff
@@ -290,6 +292,23 @@ namespace sil_block {
+  using SILDebugScopeLayout = BCRecordLayout<
+    SIL_DEBUG_SCOPE,
+    BCFixed<1>,
+    ValueIDField, // Parent
```
```suggestion
    ValueIDField, ///< Parent.
```

adrian-prantl commented on lib/Serialization/SerializeSIL.cpp:
```diff
@@ -3025,6 +3051,74 @@ void SILSerializer::writeSILProperty(const SILProperty &prop) {
     componentValues);
 }
 
+// Serialize and write SILDebugScope graph in post order
```
this should be a doxygen comment and in the header file

adrian-prantl commented on lib/Serialization/SerializeSIL.cpp:
```diff
@@ -3025,6 +3051,74 @@ void SILSerializer::writeSILProperty(const SILProperty &prop) {
+                                     const SourceManager &SM) {
+
+  if (DebugScopeMap.find(Scope) != DebugScopeMap.end()) {
+    // We won't be in a recursive call here
```
```suggestion
    // We won't be in a recursive call here.
```

adrian-prantl commented on lib/Serialization/SerializeSIL.cpp:
```diff
@@ -3025,6 +3051,74 @@ void SILSerializer::writeSILProperty(const SILProperty &prop) {
+  unsigned isFuncParent = 0;
+
+  assert(!Scope->Parent.isNull());
+  if (Scope->Parent) {
```
since there is an assert, should this return an Error if not and early exit?

adrian-prantl commented on lib/Serialization/SerializeSIL.cpp:
```diff
@@ -3025,6 +3051,74 @@ void SILSerializer::writeSILProperty(const SILProperty &prop) {
+        writeDebugScopes(Parent, SM);
+      }
+      ParentID = DebugScopeMap[Parent];
+    } else {
```
It might not hurt to comment what these two cases mean at a high level.

usama54321 committed:
Add more test cases for transparent/alwaysEmitIntoClient. Remove scopes
deserialized from stdlib

usama54321 commented:
@swift-ci test

usama54321 committed:
Remove full paths from test

usama54321 committed:
Test dump input on fail

usama54321 commented:
@swift-ci please test windows platform

usama54321 commented:
@swift-ci please test windows platform

usama54321 commented:
@swift-ci please test windows platform

usama54321 commented:
@swift-ci please test

usama54321 commented:
@swift-ci test

usama54321 committed:
Additional comments about the different orders of deserialization and
bugfixes

usama54321 commented:
@swift-ci test

adrian-prantl commented on lib/SIL/Parser/ParseSIL.cpp:
```diff
@@ -125,6 +130,11 @@ ParseSILModuleRequest::evaluate(Evaluator &evaluator,
     return SILModule::createEmptyModule(desc.context, desc.conv, desc.opts);
   }
+
+  //mark functions as zombies before calling SILVerifier as functions referred
```
```suggestion
  // Mark functions as zombies before calling SILVerifier as functions referred
  //  to by debug scopes only can fail verifier checks.
```

usama54321 commented on lib/Serialization/DeserializeSIL.cpp:
```diff
@@ -628,12 +630,29 @@ llvm::Expected<SILFunction *> SILDeserializer::readSILFunctionChecked(
+    //  ii) Deserialized by linker/optimizer -> deserialized for debugging ->
+    //  no zombie created
+
+    if (fn->isZombie() && !forDebugScope) {
```
@eeckstein I hope you can take a look at this again. Thanks

eeckstein replied:
lgtm

eeckstein reviewed (approved):
lgtm!

usama54321 committed:
Initial commit for Debug Scope serialization/deserialization

This patch adds code for serializing and deserializing debug scopes.
During serialization, declarations of functions only referred to by
debug scopes are serialization. During deserialization, these
declarations are deserialized as zombie functions. The deserializer
refuses to deserialize these functions when referred to by the
optimizer, such as during Generic Specialization.

usama54321 committed:
Cosmetic fixes and do not abort program if fail to deserialize debug
scope

usama54321 committed:
Enable debug info serialization by default

usama54321 committed:
Revert "Enable debug info serialization by default"

This reverts commit 384877fc1dffc3ed8aa7a7c148808106d9f9e39b.

usama54321 committed:
Fix comment

usama54321 commented:
@swift-ci test

usama54321 commented:
@swift-ci please test macOS platform
